src/compas/files/gltf/gltf_mesh.py

```diff
@@ -0,0 +1,185 @@
+    get_mode(list faces)
+        Returns the glTF mode of a list of faces.
+    validate_faces(list faces)
+        Raises an exception if not all faces in `faces` are defining either all triangles, lines
+        or points.
+    validate_vertices(list vertices)
+        Raise an exception if there are either too many vertices, or the vertices do not
+        represent points in 3-space.
+    from_vertices_and_faces(GLTFContent context, list vertices, list faces, str mesh_name, object extras)
+        Construct a :class:`compas.files.GLTFMesh` object from lists of vertices and faces.
+        Vertices can be given as either a list of xyz-tuples or -lists, in which case
+        the faces reference vertices by index, or vertices can be given as a dictionary of
+        key-value pairs where the values are xyz-tuples or -lists and the faces reference the keys.
+    from_mesh(Mesh mesh)
+        Construct a :class:`compas.files.MeshData` object from a :class:`compas.datastructures.Mesh`.
+    to_dict(list primitives)
+        Returns a JSONable dictionary object in accordance with glTF specifications.
+    from_dict(dict mesh, GLTFContent context, list primitive_data_list)
+        Creates a :class:`compas.files.GLTFMesh` from a glTF node dictionary
+        and inserts it in the provided context.
+    """
+    def __init__(self, primitive_data_list, context, mesh_name=None, weights=None, extras=None, extensions=None):
+        self.mesh_name = mesh_name
+        self.weights = weights
+        self.primitive_data_list = primitive_data_list
+        self.extras = extras
+        self.extensions = extensions
+
+        self._key = None
+        self.context = context
+        self._set_key()
+
+    def _set_key(self):
+        key = len(self.context.meshes)
+        while key in self.context.meshes:
+            key += 1
+        self.context.meshes[key] = self
+        self._key = key
+
+    @property
+    def key(self):
+        return self._key
+
+    @property
+    def vertices(self):
+        if not self.weights:
+            return list(itertools.chain(*[primitive.attributes['POSITION'] for primitive in self.primitive_data_list]))
+        return get_weighted_mesh_vertices(self, self.weights)
+
+    @property
+    def faces(self):
+        faces = []
+        shift = 0
+        for primitive_data in self.primitive_data_list:
+            shifted_indices = self.shift_indices(primitive_data.indices, shift)
+            group_size = VERTEX_COUNT_BY_MODE[primitive_data.mode]
+            grouped_indices = self.group_indices(shifted_indices, group_size)
+            faces.extend(grouped_indices)
+            shift += len(primitive_data.attributes['POSITION'])
+        return faces
+
+    def shift_indices(self, indices, shift):
+        return [index + shift for index in indices]
+
+    def group_indices(self, indices, group_size):
+        it = [iter(indices)] * group_size
+        return list(zip(*it))
+
+    @classmethod
```

Comment:
just out of curiosity, why is this not `@staticmethod`?

Reply:
Because I have used it in the `classmethod` `from_vertices_and_faces`.

Reply:
so it is a helper method for one of the other `classmethod`s?

Reply:
at the moment, yes, but i am rethinking that.  suggestions?

Reply:
is it important that they are part of the class definition? they don't seem to be using any of the attributes or methods related to the class. so perhaps they can just be helper functions?

Reply:
in any case, i think you can also just mark them as static which would already get rid of this unused and implicit `cls` argument...

Reply:
Yes, I will make it a helper method.